apis/condition_set.go

```diff
@@ -260,8 +260,8 @@ func (r conditionsImpl) MarkTrue(t ConditionType) {
 		if !c.IsTrue() {
 			// Update the happy condition if the current ready condition is
 			// marked not ready because of this condition.
-			if org.Reason == orgTL.Reason && org.Message == orgTL.Message {
-				r.propagateFailure(org, orgTL)
+			if org.GetReason() == orgTL.GetReason() && org.GetMessage() == orgTL.GetMessage() {
```

Comment:
`org` and `orgTL` only exist for this predicate.  Instead of copying on 246/247, check this condition there.  Is would also be significantly clearer to nil check the condition there vs. adding nil-sensitive getters.

I'd also structure this differently.  Rather than adding `propagateFailure`, I would create a method for finding the new `*apis.Condition` that should be reflected in `Ready` (extracted from `propagateFailure`).  Why a pointer?  `nil` means do the SetCondition (below) and non-nil means SetCondition to the returned Condition.

Reply:
but how do you know when the ready reason should be cleared if you don't capture what they were before setting the t value?